MarkBind/markbind
# Add background build option to one-page serve

**What is the purpose of this pull request?**


One of the enhancements mentioned on #1405, waiting for #1538 and #1545 to merge first



**Overview of changes:**
Adds a new `-b` option to `markbind serve`, which if used in conjunction with `-o`, will enhance the serve process by building pages that are not yet built or marked to rebuild in the background.

**_Build Interrupts_**

As background building can take some time, interrupts are necessary to stop the background build so that changes in source files can still be rebuilt in real-time. Whenever user edits the source files, background building will be stopped, affected changes will be rebuilt first, then it will resume with another instance of background build that builds the remaining pages.

This is implemented by `stopBuildTimeThreshold` property in `Site.pageGenerationContext` object, which holds the latest time that any page generation operations is permitted to run (i.e. any operations below the threshold will have to be terminated early).

Throughout the page generation pipeline (`runPageGenerationTasks` -> `generatePagesSequential`/`generatePagesAsyncThrottled`), checks to this property is done before the page is to be generated (or in the case on asynchronous processing, before the current page is to be generated and before the next page generation is issued to run).

The property will only be updated through `Site.stopOngoingBuilds()`, which will be called whenever fs events are detected.

**_Search Data Population_**

One of background-build goals is to gradually populate search data as pages get built, so users can easily navigate while being in development to remedy the shortcoming of `-o`. This is done by issuing `Site.writeSiteData` after every page build so `siteData.json` is kept in sync with the pages built as the background build goes on. As the function is called numerous of times, I had to add a parameter to toggle the logs in that function.

Fetching `siteData.json` as it gets updated, either through live-reloads or explicit fetch, found out to be not ideal as that means each opened page has to fetch the file every time a page is built, so the number of fetches scales up to the number of pages in the background build, which is not suitable for medium-sized websites with many pages.

The alternative then is to fetch the `siteData.json` after background build is finished. All tabs will be explicitly reloaded such that they all retrieve the complete `siteData.json`. Even so, users visiting a page during background build will fetch `siteData.json`, which because of the update after every page build, includes all pages that are built up to that point in time in the search data.

**Anything you'd like to highlight / discuss:**
Feedback on the general approach is appreciated.

**Testing instructions:**
Try testing out using `markbind serve -o -b` with different cases:
- Staying still, letting the background build run its course
- Try to disturb the background build by editing files and see whether the stop-then-resume flow runs properly (just Ctrl+S the files without any changes is enough to trigger fs events)
- System responsiveness test: try doing anything else during the background build



**Proposed commit message: (wrap lines at 72 characters)**
Add background build option to one-page serve

Running `markbind serve -o` only builds pages as user visits
those pages. As search data is populated according to built pages,
it is then limited to the pages that users have visited before,
rendering the search bar ineffective in live preview. Also, as
`serve -o` builds pages on-demand, the time between page builds are not
effectively utilized to build other pages.

Let's add an option to `serve` that can utilize the time after the
landing page has been generated to do a background build of pages that
are not yet built, updating the search data as it goes so users can get
not just the pages that they visited but also the pages that are built
in the background.



---

**Checklist:** :ballot_box_with_check:
